framework/src/kernels/ADKernel.C

```diff
@@ -138,36 +138,63 @@ ADKernelTempl<T>::computeResidual()
 
 template <typename T>
 void
-ADKernelTempl<T>::computeJacobian()
+ADKernelTempl<T>::computeResidualsForJacobian()
 {
-  prepareMatrixTag(_assembly, _var.number(), _var.number());
-
-  auto ad_offset =
-      Moose::adOffset(_var.number(), _sys.getMaxVarNDofsPerElem(), Moose::ElementType::Element);
+  if (_residuals.size() != _test.size())
+    _residuals.resize(_test.size(), 0);
+  for (auto & r : _residuals)
+    r = 0;
 
   precalculateResidual();
-
   if (_use_displaced_mesh)
-    for (_i = 0; _i < _test.size(); _i++)
-      for (_qp = 0; _qp < _qrule->n_points(); _qp++)
-      {
-        DualReal residual = _ad_JxW[_qp] * _ad_coord[_qp] * computeQpResidual();
-        for (_j = 0; _j < _var.phiSize(); _j++)
-          _local_ke(_i, _j) += residual.derivatives()[ad_offset + _j];
-      }
+    for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+    {
+      _r = _ad_JxW[_qp];
+      _r *= _ad_coord[_qp];
+      for (_i = 0; _i < _test.size(); _i++)
+        _residuals[_i] += _r * computeQpResidual();
+    }
```

Comment:
What is the purpose for collecting the separated residual contribs here instead of just dumping them into local_ke like before?

Reply:
Ah - I see - this is how you add the per-entry scaling - which wouldn't work before when they were all munged together.  Is there anywhere in doc comments or commit messages that explains this necessity/reason?  That might be worth capturing for future poor souls who need to dig around in this code.

Reply:
This is somewhat related to global indexing, but this code was like this well before. Even for non-global you want to save off the residuals because you can use those residuals to fill `_local_ke` for every associated jvariable. We do not want to re-do these calculations for every jvariable and corresponding `_local_ke`

Reply:
I added more doxygen to `computeResidualsForJacobian` in `ADKernel.h` and `ADIntegratedBC.h` indicating the purpose of the method